docs/middleware.md

```diff
@@ -232,6 +232,271 @@ around explicitly, rather than mutating the middleware instance.
+from asgiref.typing import ASGIReceiveEvent, ASGISendEvent
+
+
+class ASGIMiddleware:
+    def __init__(self, app: ASGI3Application) -> None:
+        self.app = app
+
+    async def __call__(self, scope: Scope, receive: ASGIReceiveCallable, send: ASGISendCallable) -> None:
+        if scope["type"] == "http":
+            async def send_wrapper(message: ASGISendCallable) -> None:
+                await send(message)
+            return await self.app(scope, receive, send_wrapper)
+        await self.app(scope, receive, send)
+```
+
+The `ASGI3Application` is meant to represent an ASGI application that follows the third version of the standard.
+Starlette itself is an ASGI 3 application.
+
+!!! note
+    You can read more about ASGI versions on the [Legacy Applications section on the ASGI documentation](https://asgi.readthedocs.io/en/latest/specs/main.html#legacy-applications).
+
+### Reusing Starlette components
+
+If you need to work with request or response data, you may find it more convenient to reuse Starlette data structures (`Request`, `Headers`, `QueryParams`, `URL`, etc) rather than work with raw ASGI data. All these components can be built from the ASGI `scope`, `receive` and `send`, allowing you to work on pure ASGI middleware at a higher level of abstraction.
+
+For example, we can create a `Request` object, and work with it.
+```python
+from starlette.requests import Request
+
+class ASGIMiddleware:
+    def __init__(self, app):
+        self.app = app
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] == "http":
+            request = Request(scope, receive, send)
+            # Do something here!
+        await self.app(scope, receive, send)
+```
+
+Or we might use `MutableHeaders` to change the response headers:
+
+```python
+class ExtraResponseHeadersMiddleware:
+    def __init__(self, app, headers):
+        self.app = app
+        self.headers = headers
+
+    async def __call__(self, scope, receive, send):
+        if scope["type"] != "http":
+            return await self.app(scope, receive, send)
+
+        async def wrapped_send(message):
+            if message["type"] == "http.response.start":
+                headers = MutableHeaders(scope=message)
+                for key, value for self.headers:
+                    headers.append(key, value)
+            await send(message)
+
+        await self.app(scope, receive, wrapped_send)
+```
+
+### Per-request state
+
+ASGI middleware classes should be stateless, as we typically don't want to leak state across requests.
+
+The risk is low when defining wrappers inside `__call__`, as state would typically be defined as inline variables.
+
+But if the middleware grows larger and more complex, you might be tempted to refactor wrappers as methods. Still, state should not be stored in the middleware instance. Instead, if you need to manipulate per-request state, you may write a separate `Responder` class:
```

Comment:
That’s wild!

Also, hmm… I’m less convinced about documenting the « responder class » now, although I’m aware I’m the one who brought it up initially. That’s because that’s a style I’ve been using. I think I was itches by the fact we’d be redefining a function every time with the closure style.

There’s a couple of styles that could be used, eg. responder class, or this generator style, etc. I think this section is complex and new enough for most readers already that we should stick to just one.

I’m curious what you would think about dropping this section. We’ve got a note somewhere on the fact that middleware classes should be stateless. The closure style actually doesn’t incite users to use class attributes, so it doesn’t pose a « per request state » problem, so we could drop this entire section.

Reply:
> That’s wild!

It is a neat pattern huh! Generators are great little state machines that you can ratchet forwards and pass messages to. And of course you can go full functional:

```python
def functional_middleware(app: ASGIApp) -> ASGIApp:
    async def new_app(scope: Scope, receive: Receive, send: Send) -> None:
        async def tweaked_send() -> "AsyncGenerator[None, Message]":
            message = yield
            headers = Headers(raw=message["headers"])
            should_tweak = headers.get("X-Tweak") == "1"
            await send(message)

            while True:
                message = yield
                if not should_tweak:
                    await send(message)
                else:
                    await send(message)

        async with aclosing(tweaked_send()) as wrapped_send:
            await wrapped_send.__anext__()  # start the generator
            await app(scope, receive, wrapped_send.asend)

    return new_app
```

> I’m curious what you would think about dropping this section.

I do think this section is getting a bit in depth, maybe at this point we should let users fly and use their imagination and dev skills to figure out which solution they want to use? I do think explaining _why_ they should not use attributes on the middleware is helpful.

Reply:
I had to re read this section several times, I think it would be extremely clear to have a `don't do this example` and another one following, `do this instead`

Reply:
I'm going to simplify this section and apply @euri10 's comment. I may remove the snippets here, and point to external implementations...

Reply:
Maybe we can even point to our own internal implementations? GZipMiddleware would be a good example.

Reply:
Hehe, I always struggle with generators. :sweat_smile: What I don't like about `yield` for receiving stuff is that there's no way to have type inference, any values obtained at the left side would have to be manually annotated to have typing support (autocompletion, inline errors, mypy, etc). I'm also not sure what better approach is there. But anyway...

I would leave this PR without more examples, to try and limit the scope. Maybe a future PR could include suggesting more examples, but I think it would be good to have this close to what is already there, without adding too much.

I would just add another phrase clarifying the stateless stuff. I'm adding that as a separate suggestion below. :nerd_face: